Fixed #31335 -- Fixed removing composed Meta constraints/indexes on foreign keys on MySQL.

GeyseR opened the pull request:
[ticket-31335](https://code.djangoproject.com/ticket/31335)

felixxm commented on tests/schema/tests.py:
```diff
@@ -2193,28 +2194,53 @@ def test_unique_together_with_fk(self):
+        with editor_with_post_cleanup('indexes') as editor:
+            editor.add_index(Book, index)
+            editor.remove_index(Book, index)
+
```
Can we add test for renaming a `ForeignKey` included in the composed index?

GeyseR replied:
Should we test that actually? As stated in the ticket, for the field renaming Django adds 3 operations to the migration file: remove an index, rename the field, add the index (back). I believe it should be at least a separate ticket, that changes behavior to one operation instead of three...

felixxm reviewed (commented):
@GeyseR Thanks for this patch :+1: I left initial comments, please also check the [comments](https://github.com/django/django/pull/12687#discussion_r405727430) from the previous PR.

GeyseR commented:
hi @felixxm, I've added two commits to the PR.
The first one (https://github.com/django/django/pull/15254/commits/dc79f8728763fde4ba38345686f778a1458c7c6d) addresses your comments in this PR.
With the second one (https://github.com/django/django/pull/15254/commits/3b029048762759d2732e424ee3bca2947fad4255) I try to apply a smarter logic for checking already existing various types of indexes, do they cover the FK index or not. In this commit, I also moved tests to `tests/migrations` as it is easier to apply migration operations with both state and DB updates instead of emulating the same behavior in the schema editor tests. For sure, these changes add complexity to the solution, but it behaves in a more optimized manner.

GeyseR commented:
hey @felixxm,
could you please help me with debugging failed MySQL CI tests?
they successfully passed on my local MySQL DB v8 and with official `django-docker-box` docker image.
Here is the output from the latter:
```
$ export MYSQL_VERSION=8
$ export DJANGO_PATH=<local django path>
$ export PYTHON_VERSION=3.10
% docker-compose run --rm mysql migrations
Creating django-docker-box_mysql_run ... done
wait-for-it.sh: waiting 20 seconds for mysql-db:3306
wait-for-it.sh: mysql-db:3306 is available after 0 seconds
Testing against Django installed in '/tests/django/django' with up to 6 processes
Found 630 test(s).
.....
Ran 630 tests in 22.102s

OK (skipped=3)
```

Do I use the correct MySQL version that matches the version on the CI server? Do you see any other reasons for the test's failure?

felixxm commented:
> Do I use the correct MySQL version that matches the version on the CI server? Do you see any other reasons for the test's failure?

CI uses MySQL 5.7 (see [the wiki page](https://code.djangoproject.com/wiki/Jenkins)). This can be an issue with caching docker images, try to remove images before switching to MySQL 5.7, e.g.
```
$ docker-compose stop
$ docker-compose rm
$ export MYSQL_VERSION=5.7
$ docker-compose run --rm mysql migrations
```

GeyseR commented:
thanks @felixxm, that helped. At least now tests fail locally :)

GeyseR commented:
I think the PR is ready to review. The single failed test looks like an unstable one and is not related to the proposed changes.

GeyseR commented:
@felixxm did you have a chance to review the updated PR

jacobtylerwalls commented:
Hi @GeyseR, if you've accounted for all feedback to date you should uncheck "Needs tests" and "Patch needs improvement" on the ticket so that there's an indication to re-review it. Rebasing from `main` would help also. Thanks.

GeyseR commented:
@jacobtylerwalls thanks for the heads up, done

felixxm commented on tests/migrations/test_operations.py:
```diff
@@ -4812,6 +4814,209 @@ def assertModelsAndTables(after_db):
+            )
+
+            actions_affect_indexes = [
+                lambda: add_remove_unique_together(["author", "title"]),
```
This case works without the patch and it looks that it's already covered by `test_alter_unique_together()`. I'd remove it.

felixxm commented on tests/migrations/test_operations.py:
```diff
@@ -4812,6 +4814,209 @@ def assertModelsAndTables(after_db):
+
+            actions_affect_indexes = [
+                lambda: add_remove_unique_together(["author", "title"]),
+                lambda: add_remove_index_together(["author", "title"]),
```
This case works without the patch and it looks that it's already covered by `test_alter_index_together()`. I'd remove it.

felixxm reviewed (commented):
@GeyseR Thanks for updates :+1:

GeyseR commented on tests/migrations/test_operations.py:
```diff
@@ -4812,6 +4814,209 @@ def assertModelsAndTables(after_db):
+                with assert_same_indexes():
+                    action()
+            execute(migrations.AlterIndexTogether("Book", []))
+            recreate_author_field()
```
year, retruning back to the PR after 3 months, I see how it could be difficult to understand what's going on there. I'll try to simplify the code.

felixxm commented:
> hey @felixxm, I tried to simplify the tests code, how does it look now?

Many thanks :rocket: I have a week off, so I'm going to review it after the holidays (June, 13th+).

GeyseR commented:
no problem, anyway I need to fix the failed tests :)

David-Wobrock commented on django/db/backends/mysql/schema.py:
```diff
@@ -135,16 +155,71 @@ def _field_should_be_indexed(self, model, field):
+        def get_first_index_field_name(index_fields, index_expressions):
+            field = None
+            if index_fields:
+                if (
+                    index_fields[0].startswith("-")
+                    and not self.connection.features.supports_index_column_ordering
+                ):
+                    field = index_fields[0][1:]
+                elif not index_fields[0].startswith("-"):
+                    field = index_fields[0]
```
If I understand correctly, when the first field is indexed in descending order (like `-fieldname`) and the DB supports it, we will return `None` for this function.
1/ MySQL doesn't drop the implicit FK index when another index has as first field the FK in descending order, is that correct?
2/ The naming of the function is not _entirely_ accurate with respect to the behaviour, since we'll return None even though one would expect to get the first field name.

GeyseR commented:
hi @David-Wobrock, thanks for the review. After some thinking I decided to change how I check for duplicate FK index: instead of checking model's definition I check the DB table structure. It looks like is much more straightforward way and also more reliable as we can find some unmanaged manual indexes that could be enough for index deletion.

After that we don't need to extract get_first_index_field_name as a helper function anymore :)

GeyseR commented on django/db/backends/mysql/schema.py:
```diff
@@ -135,16 +155,71 @@ def _field_should_be_indexed(self, model, field):
+                and LOOKUP_SEP not in index_expressions[0].name
+            ):
+                field = expressions[0].name
+            return field
```
get_first_index_field_name is completely dropped (see the comment in the main thread)

GeyseR commented on django/db/backends/mysql/schema.py:
```diff
@@ -135,16 +155,71 @@ def _field_should_be_indexed(self, model, field):
+        self,
+        model,
+        *,
+        fields=None,
```
good catch, fixed

GeyseR commented on tests/schema/tests.py:
```diff
@@ -2527,46 +2527,6 @@ def test_unique_together(self):
 
-    def test_unique_together_with_fk(self):
-        """
-        Tests removing and adding unique_together constraints that include
-        a foreign key.
-        """
-        # Create the table
-        with connection.schema_editor() as editor:
-            editor.create_model(Author)
-            editor.create_model(Book)
-        # Ensure the fields are unique to begin with
-        self.assertEqual(Book._meta.unique_together, ())
-        # Add the unique_together constraint
-        with connection.schema_editor() as editor:
-            editor.alter_unique_together(Book, [], [["author", "title"]])
-        # Alter it back
-        with connection.schema_editor() as editor:
-            editor.alter_unique_together(Book, [["author", "title"]], [])
-
-    def test_unique_together_with_fk_with_existing_index(self):
-        """
-        Tests removing and adding unique_together constraints that include
-        a foreign key, where the foreign key is added after the model is
-        created.
-        """
-        # Create the tables
-        with connection.schema_editor() as editor:
-            editor.create_model(Author)
-            editor.create_model(BookWithoutAuthor)
-            new_field = ForeignKey(Author, CASCADE)
-            new_field.set_attributes_from_name("author")
-            editor.add_field(BookWithoutAuthor, new_field)
-        # Ensure the fields aren't unique to begin with
-        self.assertEqual(Book._meta.unique_together, ())
-        # Add the unique_together constraint
-        with connection.schema_editor() as editor:
-            editor.alter_unique_together(Book, [], [["author", "title"]])
-        # Alter it back
-        with connection.schema_editor() as editor:
-            editor.alter_unique_together(Book, [["author", "title"]], [])
```
the cases that are tested by these two tests are also covered by the new tests in the `migrations.test_operations.OperationTests` class. Do you think we should keep these instead and drop part of new code (I would prefer to have all tests grouped and located in one place)

David-Wobrock replied:
Ah okay yes, I was mislead since I didn't read through the added unit tests. My bad 🤕

felixxm replied:
This is an issue in `schema` so I'd keep these tests in `schema/tests.py` and also add new tests here.

David-Wobrock commented on tests/migrations/test_operations.py:
```diff
@@ -5182,6 +5182,319 @@ def assertModelsAndTables(after_db):
+            else:
+                self.assertEqual(set(actual_indexes), set(initial_indexes))
+
+        assert_index_created(
+            [
+                migrations.AlterUniqueTogether("Book", ["author", "title"]),
+                migrations.AlterUniqueTogether("Book", []),
+            ]
+        )
+        assert_index_created(
+            [
+                migrations.AlterIndexTogether("Book", ["author", "title"]),
+                migrations.AlterIndexTogether("Book", []),
+            ]
+        )
```
Twice the same. Did you mean
```suggestion
        assert_index_created(
            [
                migrations.AlterUniqueTogether("Book", ["author", "title"]),
                migrations.AlterUniqueTogether("Book", []),
            ]
        )
        assert_index_created(
            [
                migrations.AlterIndexTogether("Book", []),
                migrations.AlterIndexTogether("Book", ["author", "title"]),
            ]
        )
```
?

GeyseR replied:
there are two different operations: AlterUniqueTogether and AlterIndexTogether.
For both we test creating composed index with a FK and then dropping it, to check that both operations behave without failure

David-Wobrock commented on tests/migrations/test_operations.py:
```diff
@@ -5182,6 +5182,319 @@ def assertModelsAndTables(after_db):
+        # combinations of operations that would create an additional index
+        # here we shouldn't do that as another index exists and includes the FK field
+        operations_to_check = [
+            [
+                migrations.AlterUniqueTogether("Book", ["author", "title"]),
+                migrations.AlterUniqueTogether("Book", []),
+            ],
+            [
+                migrations.AlterIndexTogether("Book", ["author", "title"]),
+                migrations.AlterIndexTogether("Book", []),
+            ],
```
Twice the same.
Did you mean:
```suggestion
            [
                migrations.AlterUniqueTogether("Book", ["author", "title"]),
                migrations.AlterUniqueTogether("Book", []),
            ],
            [
                migrations.AlterIndexTogether("Book", []),
                migrations.AlterIndexTogether("Book", ["author", "title"]),
            ],
```
?

David-Wobrock reviewed (commented):
In a more subjective matter, I find the unit tests with their nested functions a bit hard to read. But since I don't have anything better to suggest at the moment, that's fine to me :)

GeyseR commented:
@felixxm the PR is ready to review 🙄 
Should I squash all commits in one?

felixxm commented:
Thanks :+1: 

> @felixxm the PR is ready to review roll_eyes Should I squash all commits in one?

No need, I can squash with final edits.

felixxm commented on tests/migrations/test_operations.py:
```diff
@@ -5227,6 +5227,321 @@ def assertModelsAndTables(after_db):
+                ]
+            )
+
+    def test_composed_func_indexes_dont_create_implicit_fk_index_when_deleting(self):
```
This tests works without a patch, we should add it in a separate commit/PR.

GeyseR replied:
makes sense. I've squashed everything related to the original bugfix in one commit and then supporting of the different kinds of indexes/constraints in another one. Does this work? Not sure that we should make another PR as in this case this PR will introduce non-working code

felixxm commented on tests/migrations/test_operations.py:
```diff
@@ -5227,6 +5227,321 @@ def assertModelsAndTables(after_db):
+        assert_index_created(
+            [
+                migrations.AlterIndexTogether("Book", ["author", "title"]),
+                migrations.AlterIndexTogether("Book", []),
```
`index_together` is deprecated so we need to move its tests to a separate test methods to make it easier to remove when deprecation ends.

felixxm commented on django/db/backends/mysql/schema.py:
```diff
@@ -135,24 +153,63 @@ def _field_should_be_indexed(self, model, field):
+
+        first_field_name = None
+        if fields:
+            if (
+                fields[0].startswith("-")
+                and not self.connection.features.supports_index_column_ordering
+            ):
+                first_field_name = fields[0][1:]
+            elif not fields[0].startswith("-"):
+                first_field_name = fields[0]
```
This is required only for indexes defined in `Meta.indexes`. Should we pass list of pure fields names in `remove_index()` instead? For example:
```python
    def remove_index(self, model, index):
        self._create_missing_fk_index(
            model,
            fields=[field_name for field_name, _ in index.fields_orders],
            expressions=index.expressions,
        )
        super().remove_index(model, index)
```

felixxm commented on django/db/backends/mysql/schema.py:
```diff
@@ -135,24 +153,63 @@ def _field_should_be_indexed(self, model, field):
+        if not first_field_name:
+            return
+
+        meta = model._meta
+
+        first_field = meta.get_field(first_field_name)
```
`meta` is used only here so a temporary variable is unnecessary:
```suggestion
        first_field = model._meta.get_field(first_field_name)
```

felixxm reviewed (commented):
@GeyseR Thanks for updated :+1: We're getting closer.

GeyseR commented:
@felixxm 
regarding this one (GH, for some reason, doesn't allow replying in the discussion thread)

> This is an issue in schema so I'd keep these tests in schema/tests.py and also add new tests here

these two tests are now covered by `test_composed_indexes_create_implicit_fk_index_when_deleting`. Why would we need to keep redundant tests and spend time (and electricity :)) for running them in each build?

I've tried to move tests from operations tests to the schema tests, and this movement also requires adding utility code (apply_operations in particular) to the `SchemaTests` class. This looks like too much additional knowledge for this class: it should become aware of migrations, the project's state, etc. To avoid that, we need to operate on more low-level schema-related code (which is how I made the first version of the patch, and it looked ugly and confusing)

felixxm commented:
> these two tests are now covered by test_composed_indexes_create_implicit_fk_index_when_deleting. Why would we need to keep redundant tests and spend time (and electricity :)) for running them in each build?

I was asking for keeping these two and moving new one to `schema/tests.py`, not about creating redundant tests.

> I've tried to move tests from operations tests to the schema tests, and this movement also requires adding utility code (apply_operations in particular) to the SchemaTests class. This looks like too much additional knowledge for this class: it should become aware of migrations, the project's state, etc. To avoid that, we need to operate on more low-level schema-related code (which is how I made the first version of the patch, and it looked ugly and confusing)

I'm not sure why `apply_operations` hook is needed and what makes the new tests "ugly" and "confusing" :thinking: This bug is related with a specific model that we can create in schema tests and then perform `remove_index()/remove_constraint()` on it to see if it works. Am I missing something? For example:
```python
    def test_meta_unique_constraint_with_fk(self):
        with connection.schema_editor() as editor:
            editor.create_model(Author)
            editor.create_model(Book)
        constraint = UniqueConstraint(
            fields=["author", "title"], name="book_author_title_uniq"
        )   
        Book._meta.constraints = [constraint]
        with connection.schema_editor() as editor: 
            editor.add_constraint(Book, constraint)
        constraints = self.get_constraints(Book._meta.db_table)
        self.assertIn(constraint.name, constraints)
        Book._meta.constraints = []
        with connection.schema_editor() as editor:
            editor.remove_constraint(Book, constraint)
        constraints = self.get_constraints(Book._meta.db_table)
        self.assertNotIn(constraint.name, constraints)

    def test_meta_index_with_fk(self):
        with connection.schema_editor() as editor:
            editor.create_model(Author)
            editor.create_model(Book)
        index = Index(
            fields=["author", "title"], name="book_author_title_idx"
        )
        Book._meta.indexes = [index]
        with connection.schema_editor() as editor:
            editor.add_index(Book, index)
        indexes = self.get_indexes(Book._meta.db_table)
        self.assertNotIn("author_id", indexes)
        Book._meta.indexes = []
        with connection.schema_editor() as editor:
            editor.remove_index(Book, index)
        indexes = self.get_indexes(Book._meta.db_table)
        self.assertIn("author_id", indexes)
```
are proper regression tests for removing `Index()` and `UniqueConstraint()` with a foreign key.

GeyseR commented:
> I was asking for keeping these two and moving new one to schema/tests.py, not about creating redundant tests.

I mean that the new tests cover the cases from these two tests, so they will be redundant as they test the same thing.

Let me make one more attempt with schema tests, I'll try to return soon :)

GeyseR commented:
hey @felixxm, I've pushed another attempt to add tests to `tests/schema` instead of `tests/migrations`
I should take my words back about code ugliness, this time it looks good to me :)

felixxm commented on django/db/backends/mysql/schema.py:
```diff
@@ -135,24 +153,55 @@ def _field_should_be_indexed(self, model, field):
+        *,
+        fields,
+        expressions=None,
+    ):
```
I moved adding this hook to a separate PR, see #16048.

felixxm commented on tests/schema/tests.py:
```diff
@@ -2680,23 +2680,150 @@ def test_unique_together(self):
+                        (editor.remove_constraint, (Book, check_constraint)),
+                    ],
+                    created=False,
+                )
```
Thanks for moving tests to the `schema`, however I still find them complicated and difficult to understand due to the number of internal hooks, flags, and parameters. They are also fragile to isolation issues as you've put them all in one test method.

This patch only fixes a MySQL-specific caveat so complicated tests are unnecessary. I would leave existing tests untouched and add sth like this:
```python
    def test_meta_unique_constraint_with_fk(self):
        with connection.schema_editor() as editor:
            editor.create_model(Author)
            editor.create_model(Book)
        constraint = UniqueConstraint(
            fields=["author", "title"], name="book_author_title_uniq"
        )
        Book._meta.constraints = [constraint]
        with connection.schema_editor() as editor:
            editor.add_constraint(Book, constraint)
        constraints = self.get_constraints(Book._meta.db_table)
        self.assertIn(constraint.name, constraints)
        Book._meta.constraints = []
        with connection.schema_editor() as editor:
            editor.remove_constraint(Book, constraint)
        constraints = self.get_constraints(Book._meta.db_table)
        self.assertNotIn(constraint.name, constraints)

    def _test_meta_index_with_fk(self, index, removed=True):
        with connection.schema_editor() as editor:
            editor.create_model(Author)
            editor.create_model(Book)
        table = Book._meta.db_table
        Book._meta.indexes = [index]
        with connection.schema_editor() as editor:
            editor.add_index(Book, index)
        self.assertIn(index.name, self.get_constraints(Book._meta.db_table))
        Book._meta.indexes = []
        with connection.schema_editor() as editor:
            editor.remove_index(Book, index)
        indexes = self.get_indexes(Book._meta.db_table)
        self.assertNotIn(index.name, self.get_constraints(Book._meta.db_table))
        
    def test_meta_index_with_fk(self):
        index = Index(fields=["author", "title"], name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    def test_meta_index_with_fk(self):
        index = Index(fields=["-author", "title"], name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    @skipUnlessDBFeature("supports_expression_indexes")
    def test_meta_func_index_with_fk(self):
        index = Index(F("author"), F("title"), name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    @skipUnlessDBFeature("supports_expression_indexes")
    def test_meta_desc_func_index_with_fk(self):
        index = Index(F("author").desc(), F("title"), name="book_author_title_idx")
        self._test_meta_index_with_fk(index)
        
    @skipUnlessDBFeature("supports_expression_indexes")
    def test_meta_func_transform_index_with_fk(self):
        index = Index(F("title__lower"), name="book_title_lower_idx")
        with register_lookup(CharField, Lower):
            self._test_meta_index_with_fk(index, removed=False)
```

GeyseR replied:
> This patch only fixes a MySQL-specific caveat so complicated tests are unnecessary.

Do you mean that we shouldn't check the implicit index existence, just check that removing indexes/constraints work?

GeyseR replied:
makes sense, I didn't known that it is verified as a bug in MySQL (I thought they consider it as a feature :) )

GeyseR replied:
done

felixxm commented:
I rebased after ec13e801b820614ff374cb0046092caab8d67249 and squashed commits.

felixxm commented on tests/schema/tests.py:
```diff
@@ -2698,6 +2698,72 @@ def test_unique_together_with_fk(self):
+    @skipUnlessDBFeature("supports_table_check_constraints")
+    def test_composed_check_constraint_with_fk(self):
+        constraint = CheckConstraint(check=Q(author__gt=0), name="book_author_check")
+        self._test_composed_constraint_with_fk(constraint)
```
I moved extra tests to a separate PR, see #16049.

felixxm commented:
@GeyseR Thanks :+1: It's ready :rocket:

GeyseR commented:
@felixxm thanks for you patience with reviewing my code 😅
the last updates look goot to me 👍
